libs/core/langchain_core/tools/retriever.py

```diff
@@ -41,12 +46,18 @@ async def _aget_relevant_documents(
     document_prompt: BasePromptTemplate,
     document_separator: str,
     callbacks: Callbacks = None,
-) -> str:
+    response_format: Literal["content", "content_and_artifact"] = "content",
+) -> Union[str, tuple[str, list[dict[str, Any]]]]:
     docs = await retriever.ainvoke(query, config={"callbacks": callbacks})
-    return document_separator.join(
+    content = document_separator.join(
         [await aformat_document(doc, document_prompt) for doc in docs]
     )
 
+    if response_format == "content_and_artifact":
+        return (content, [doc.model_dump() for doc in docs])
```

Comment:
might make sense to just return `docs` as the artifact - any reason to dump here in particular?

Reply:
The reason is to always have the same type after deserialization.

As `artifact` has type `Any` when deserializing a `ToolMessage` from json or any other source (api call, DB, etc), Pydantic won't be able to do any "magic" to regenerate the artifact as a `list[Document]` but it will be a plain `list[dict]`.

So, to be consistent and always have a dict I opted to dump.
Otherwise you need to add custom logic when deserializing to have consistent types.